design-proposals/klip-42-schema-migrations-tool.md

```diff
@@ -0,0 +1,442 @@
+
+The migration files follow the same `Flyway` naming convention.
+`(Prefix)(Version)__(Description).sql`
+
+The `Prefix` specifies whether the file is a new migration (`V`) or undo (`U`) file.
+The `Version` is the version number used for the schema. For minor versions, such as `1.1`, an underscore is required (i.e. `1_1`)
+The `Description` is just a name or description of the new migration.
+
+i.e.
+```
+- V1__Initial_setup.sql    # a new version to migrate (v1)
+- U1__Initial_setup.sql    # rollback v1 schema
+- V2__Add_users.sql        # a new version to migrate (v2)
+- U2__Add_users.sql        # rollback v2 schema
+- V2_1__Fix_topic_name.sql # A new version to migrate (v2.1)
+- U2_1__Fix_topic_name.sql # rollback v2.1 schema
+```
+
+It is recommended the user creates these two files on any new migration. The tool will not enforce that. We need specify this recomendation in the ksqlDB documents.
+
+### Command Syntax
+
+Finally, let's look at the rest of the command parameters that will exist to facilitate schema migrations.
+
+```shell script
+Usage:
+  ksql-migrations [options] commands
+  
+Commands
+  initialize   Initializes the schema version table
+
+  apply ( all | next | until <target> )
+  
+              Migrates a schema to new available schema versions (default: next)
+              
+              If 'all' is specified, then it migrates all newer versions available
+              If 'next' is specified, then it migrates only the next available version
+              If 'until <target>' is specified, then it migrates all available versions before <target>
+  
+  undo ( all | last | until <target> )
+  
+              Rollbacks a schema to the previous schema version (default: last)
+              
+              If 'all' is specified, then it rollbacks all previous versions
+              If 'last' is specified, then it rollbacks the previous version
+              If 'until <target>' is specified, then it rollbacks all previous versions after <target>
+  
+  info        Displays information about the current and available migrations
+  
+  baseline    Sets the current schema to the specified version
+  
+              This is useful when production environments already have a schema unversioned. This sets
+              the baseline from where to start applying migrations (i.e. kqsl-migrations baseline 1 'Current State')
+  
+  clean       Cleans the schema metadata objects                
+  
+  validate    Validate applied migrations against local files
+  
+              Compares local files checksum against the current metadata checksums to check for migrations files that have changed.
+              This tells the user that their schema might be not valid against their local files.
+  
+Options
+  -c, --config-file  Specifies a configuration file
+ 
+  -h, --help         Shows this help  
+    
```

Comment:
How feasible would it be to support variables / `--define` as a ksql-migrate CLI option?

Let's say I want to apply the same migration in dev and prod, but the topic names I use are different between environments. If I could specify `kafka_topic='${topicName}'` in my migration, I could reuse the migration with `ksql-migrate apply -d topicName=dev-mytopic`.

Reply:
That's something we could add too. Seems useful. I'll add it to the klip.

Reply:
Done. I added the parameter to the tool help.

Reply:
+1, I think this is a great idea @colinhicks.